tests for rcl_wait

jacquelinekay opened the pull request:
I wrote a few tests for simple cases in `rcl_wait` and made a fix to the way timeouts and timers are handled in `rcl_wait`.

The cases are:

- positive timeout argument (1 ms)
- negative timeout argument (block forever) and a timer with 1 ms callback
- zero timeout (nonblocking)
- negative timeout argument with guard condition in the waitset that is triggered from another thread.

wjwwood commented on rcl/test/rcl/test_wait.cpp:
```diff
@@ -27,6 +30,8 @@
 #endif
 
 
+#define TOLERANCE 100000  // clock error tolerance in nanoseconds
```
You could use `RCL_S_TO_NS` here if you feel you need to describe it anyways.

wjwwood replied:
I guess that there isn't a precise enough conversion for this in the `rcl` macros, i.e. there is no `RCL_US_TO_NS`.

jacquelinekay replied:
https://github.com/ros2/rcl/blob/master/rcl/include/rcl/time.h#L27-L33 ?

wjwwood replied:
Ah, just kidding there is... Past self is smarter than current self apparently. Yes those macros are the one's I was talking about.

wjwwood commented on rcl/test/rcl/test_wait.cpp:
```diff
@@ -41,3 +46,132 @@ TEST(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), test_resize_to_zero) {
+  rcl_ret_t ret = rcl_wait_set_init(&wait_set, 0, 0, 1, 0, 0, rcl_get_default_allocator());
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+
+  int64_t timeout = 1000000;  // nanoseconds
```
Again I'd recommend `RCL_MS_TO_NS` or something like that.

wjwwood commented on rcl/test/rcl/test_wait.cpp:
```diff
@@ -41,3 +46,132 @@ TEST(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), test_resize_to_zero) {
   ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
 }
+
+// Some test cases for the waitset
```
Is there a more appropriate comment you can put here, one that perhaps summarizes what particular API or behaviors of the waitset are being tested?

wjwwood commented on rcl/test/rcl/test_wait.cpp:
```diff
@@ -41,3 +46,132 @@ TEST(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), test_resize_to_zero) {
+  ret = rcl_wait(&wait_set, timeout);
+  std::chrono::steady_clock::time_point after_sc = std::chrono::steady_clock::now();
+  // Check time
+  int64_t diff = (after_sc - before_sc).count();
```
Are you sure `.count` will always return nanoseconds? I think it might depend on the default storage of the clock being used (which might be implementation specific). I'd recommend doing an explicit `duration_cast` to nanoseconds first.

wjwwood commented on rcl/test/rcl/test_wait.cpp:
```diff
@@ -41,3 +46,132 @@ TEST(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), test_resize_to_zero) {
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+
+
+  // TODO(jacquelinekay) rcl timer tests?
```
More tests or is this a stale TODO?

jacquelinekay replied:
I was considering adding a new `test_timer` file that tests timers without a wait_set involved, in anticipation of seeing more regressions with rclcpp using rcl. But, because I didn't see those regressions in my other branch, I think I'm just going to take out this TODO and then add timer unit tests in the future as they become an issue. Does that sound OK?

wjwwood replied:
Fine by me.

wjwwood commented on rcl/test/rcl/test_wait.cpp:
```diff
@@ -41,3 +46,132 @@ TEST(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), test_resize_to_zero) {
+  // TODO(jacquelinekay) rcl timer tests?
+  rcl_timer_t timer = rcl_get_zero_initialized_timer();
+  rcl_timer_callback_t callback;
+  ret = rcl_timer_init(&timer, 1000000, callback, rcl_get_default_allocator());
```
Should you pass `NULL` or `nullptr` (since this is a cpp file) instead of an uninitialized `rcl_timer_callback_t` (essentially a function pointer)?

jacquelinekay replied:
hmm, yeah, I forgot that `rcl_timer_callback_t` is a typedef to a pointer type--I glanced at the header file and thought that I had to pass an object by const reference, which come to think of it doesn't make much sense...

wjwwood commented on rcl/test/rcl/test_wait.cpp:
```diff
@@ -41,3 +46,132 @@ TEST(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), test_resize_to_zero) {
+  std::chrono::steady_clock::time_point before_sc = std::chrono::steady_clock::now();
+  ret = rcl_wait(&wait_set, timeout);
+  std::chrono::steady_clock::time_point after_sc = std::chrono::steady_clock::now();
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
```
Looks like this would cause the guard condition and the wait set to memory leak, but that's probably not a big issue. Nodes and `rcl_init`/`rcl_shutdown` are more important to clean up in exception cases.

jacquelinekay replied:
yeah I mean, I could change ASSERT to EXPECT, or I could add a `fail:` block (here and in several other places). If I added a fail block I would probably generalize some of these cases better by adding custom startup/shutdown behaviors in the `WaitSetTestFixture` class. But also, I'm lazy.

wjwwood replied:
I think the common pattern is to use a "scope exit" to handle the clean up, e.g.:

https://github.com/ros2/rcl/blob/master/rcl/test/rcl/test_node.cpp#L90-L94

wjwwood commented:
lgtm other than some comments.

jacquelinekay commented:
I think I've addressed all comments (I decided to go ahead and implement scope exits here, it wasn't so bad).

I launched the CI with `--end-with rcl`. There are some failures in the services nosetest on default which I will address separately.

http://ci.ros2.org/job/ci_linux/1202/
http://ci.ros2.org/job/ci_osx/972/
http://ci.ros2.org/job/ci_windows/1251/

jacquelinekay commented:
I think maybe the tolerances are too strict for our Jenkins machines.

jacquelinekay commented:
relaunched
http://ci.ros2.org/job/ci_linux/1205/
http://ci.ros2.org/job/ci_osx/973/
http://ci.ros2.org/job/ci_windows/1252/
If these are OK I'll merge and then relaunch CI for use_rcl.
